kf/bpf.go

```diff
@@ -1098,6 +1108,110 @@ func (b *BPF) VerifyMetricsMapsVanish() error {
 	return err
 }
 
+// LoadXDPAttachProgram - Load and attach xdp root program or any xdp program when chaining is disabled
+func (b *BPF) LoadXDPAttachProgram(ifaceName string, eBPFProgram *BPF) error {
+	iface, err := net.InterfaceByName(ifaceName)
+	if err != nil {
+		log.Fatal().Msgf("lookup network iface %q: %s", ifaceName, err)
+	}
+
+	CollectionRef, err := ebpf.LoadCollection(eBPFProgram.Program.ObjectFile)
+	if err != nil {
+		return fmt.Errorf("%s: loading of xdp root failed", eBPFProgram.Program.ObjectFile)
+	}
+
+	b.ProgMapCollection = CollectionRef
+	bpfRootProg := CollectionRef.Programs[eBPFProgram.Program.EntryFunctionName]
+	bpfRootMap := CollectionRef.Maps[eBPFProgram.Program.MapName]
+
+	// Pinning root map
+	rootArrayMapFileName := filepath.Join(b.hostConfig.BpfMapDefaultPath, eBPFProgram.Program.MapName)
+	if err := bpfRootMap.Pin(rootArrayMapFileName); err != nil {
+		return fmt.Errorf("%s:failed to pin the map", rootArrayMapFileName)
+	}
+
+	_, err = link.AttachXDP(link.XDPOptions{
+		Program:   bpfRootProg,
+		Interface: iface.Index,
+	})
+	if err != nil {
+		return fmt.Errorf("could not attach XDP program: %s", err)
+	}
+
+	progInfo, err := bpfRootProg.Info()
+	if err != nil {
+		return fmt.Errorf("could not get program info of xdp root program: %s", err)
+	}
+
+	ok := false
+	b.ProgID, ok = progInfo.ID()
+	if !ok {
+		return fmt.Errorf("failed to fetch the xdp root Program ID")
+	}
+
+	ebpfInfo, err := bpfRootMap.Info()
+	if err != nil {
+		return fmt.Errorf("fetching map info failed %v", err)
+	}
+
+	b.ProgMapID, ok = ebpfInfo.ID()
+	if !ok {
+		return fmt.Errorf("fetching map id failed %v", err)
+	}
+
+	return nil
+}
+
+// UnLoadProgram - Unload the program
+func (b *BPF) UnLoadProgram(ifaceName, direction string) error {
+	_, err := net.InterfaceByName(ifaceName)
+	if err != nil {
+		log.Fatal().Msgf("lookup network iface %q: %s", ifaceName, err)
+	}
+
```

Comment:
Are we planning to support stopping user space program along with the eBPF program here?

Reply:
yes, in the next PR, when we load ebpf_program and start user program combination, will add stopping user program.

Reply:
Can we add a TODO line then?

Reply:
Done